### PR TITLE
LOG-660: Enable pod base filtering using label router plugin

### PR DESCRIPTION
**Pull request by k-keiichi-rh**
# Overview

This PR adds pod label base filtering feature based on [label router plugin](https://github.com/banzaicloud/fluent-plugin-label-router).

To implement this PR, fluentd needs pod labels information to forward application logs only from pods identified by labels.
However we thought that the fluentd config generator should be independent with the kubernetes API server as discussed in https://github.com/openshift/cluster-logging-operator/issues/787 in cluster-logging-operator.

The [label router plugin](https://github.com/banzaicloud/fluent-plugin-label-router) routes records based on their Kubernetes metadata. So this plugin could solve this problem.

### Example LF CR:
```
spec:
  inputs:
   - name: app-product-logs
     application:
       selector: "environment: production"
  outputs:
   - name: syslog
     ...
  pipelines:
   - name: forward-product-logs-to-syslog
     inputRefs:
     - app-product-logs
     outputRefs:
     - syslog
```
### fluent.conf after applying Example LF CR:
```
<label @_APPLICATION>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @FORWARD_PRODUCT_LOGS_TO_SYSLOG_BASED_ON_LABEL
    </store>
  </match>
</label>
<label @FORWARD_PRODUCT_LOGS_TO_SYSLOG_BASED_ON_LABEL>
  <match **>
    @type label_router
    <route>
      @label @FORWARD_TO_SYSLOG
      <match>
        labels environment: production
      </match>
    </route>
  </match>
</label>
<label @FORWARD_TO_SYSLOG>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @SYSLOG
    </store>
  </match>
</label>
...
```
# Tasks
## Proof of Concept implementation
Status: Completed

Refer to the details in https://github.com/openshift/cluster-logging-operator/pull/865#issuecomment-778491584

## Support multiple outputs
Status: In progress(Implemented and review required)

InputSelectorBlock as a new pipeline is introduced into ClusterLogForwarder to route records based on their Kubernetes metadata using label router plugin.
InputSelectorBlock is a single filter point to connect application logs identified by labels and namespaces with application pipelines. 
If InputSelectorBlock is not activated, it doesn't affect the existing process of generating fluent.conf.

### Example LF CR:
```
spec:
  inputs:
   - name: app-product-logs
     application:
       selector
         matchLabels: 
           environment: production"
   - name: app-dev-logs
     application:
       selector
         matchLabels: 
           environment: dev"
  outputs:
   - name: syslog1
     ...
   - name: syslog2
     ...
  pipelines:
   - name: forward-product-logs-to-syslog1
     inputRefs:
     - app-product-logs
     outputRefs:
     - syslog1
   - name: forward-dev-logs-to-syslog2
     inputRefs:
     - app-dev-logs
     outputRefs:
     - syslog2
   - name: all-default
     inputRefs:
     - audit
     - infrastructure
     - application
     outputRefs:
     - default
```

### fluent.conf after applying Example LF CR:
```
<label @_APPLICATION>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
    <store>
      @type relabel
      @label @_APPLICATION_BASED_ON_INPUT_SELECTOR
    </store>
  </match>
</label>
<label @_APPLICATION_BASED_ON_INPUT_SELECTOR>
<match **>
    @type label_router
    <route>
      @label @FORWARD_PRODUCT_LOGS_TO_SYSLOG1
      <match>
        labels environment:production
      </match>
    </route>
    <route>
      @label @FORWARD_DEV_LOGS_TO_SYSLOG2
      <match>
        labels environment:dev
      </match>
    </route>
  </match>
</label>

<label @_AUDIT>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
  </match>
</label>

<label @_INFRASTRUCTURE>
  <match **>
    @type copy
    <store>
      @type relabel
      @label @ALL_DEFAULT
    </store>
  </match>
</label>

```

### Fix namespace behavior
Status: Implemented and review required

### Test and add testcases
Status: Implemented and review required

### Support LabelSelector to identify pods
This task won't be completed in this PR.
The label router plugin doesn't support LabelSelector to route records. So we need to extend this plugin.
There is no kubernetes library for Ruby [here](https://kubernetes.io/docs/reference/using-api/client-libraries/) that supports LabelSelector like label package in [apimachinery](https://github.com/kubernetes/apimachinery). So I need to implement same feature to achieve this PR completely.

/cc @alanconway @jcantrill
/assign 

### Links

- Github issue: https://github.com/openshift/cluster-logging-operator/issues/787
- Enhancement proposal: https://github.com/openshift/enhancements/pull/457
- JIRA: https://issues.redhat.com/browse/LOG-660

### CODE REVIEW

**Comment by alanconway** [comment]
The [label_router](https://github.com/banzaicloud/fluent-plugin-label-router/blob/master/README.md) also supports namespace filtering. Our existing namespace selector is buggy, we should replace it by using the label_router for both namespace and label filtering. This should be more efficient and give simpler configuration.

For the first implementation, I think the simple `{ name: value, name: value }` equality-based selector is sufficient. We can  create another story to implement the set-based selectors. 

@jcantrill makes a good point - we already have a plugin that makes API calls to read labels and namespaces. Ideally we don't want to do that twice. It might be necessary to build a hybrid plug-in that makes API calls once and re-uses the same metadata for annotations and routing.

However this all looks very promising, once we figure out the best way to put it together. I'm particularly happy that we can incorporate the namespace selectors into the same plugin as that fixes a bug and simplifies the code at the same time.

**Comment by k-keiichi-rh** [comment]
> This looks very promising. I suggest the following steps:

Thank you for your review. I will work for this PR with the following steps.

> * Just implement the MatchLabels part of LabelSelector at first - that's easy to map to the fluentd config. Copy the LabelSelector struct and delete MatchExpressions, so we can add MatchExpressions back later. I'll modify the enhancement proposal to say that.

I agreed to implement MatchLabels part at first and copy LabelSelector struct to cluster-logging-operator.
Before moving forward my work, how's about **https://github.com/k-keiichi-rh/cluster-logging-operator/commit/23d8c46542174bd4fd8249d962df10c53ffcfd9e**?
I think CustomResourceDefinition will help us to limit the functionality of LabelSelector.
Please correct me if my understanding is wrong.

> * [label_router](https://github.com/banzaicloud/fluent-plugin-label-router/blob/master/README.md) also routes on namespaces. Replace the existing namespace code so we do both in one place. The existing code has a [bug](https://bugzilla.redhat.com/show_bug.cgi?id=1925361).  so that will fix the bug and simplify the code.

I also think so. I will fix namespace's behavior.

**Comment by alanconway** [comment]
>  Before moving forward my work, how's about **[k-keiichi-rh@23d8c46](https://github.com/k-keiichi-rh/cluster-logging-operator/commit/23d8c46542174bd4fd8249d962df10c53ffcfd9e)**?

Looks good, carry on!

**Comment by k-keiichi-rh** [comment]
I updated current status of this PR in https://github.com/openshift/cluster-logging-operator/pull/865#issue-560075225

**Review by alanconway — CHANGES_REQUESTED** [review]
Looks good but needs some tests, follow the example of existing tests for the namespace selector.

**Comment by alanconway** [comment]
/cc @vimalk78 
@vimalk78 you should take a look at this - it affect the namespace selector routing code.

**Comment by vimalk78** [comment]
Can you please add some tests

**Comment by k-keiichi-rh** [comment]
> Can you please add some tests

Thanks. I will add some unit and e2e tests by referring to the following commits:

51c3b63ecdd8aff5281c40146246f7e2287b5330
cebe0991ef8102ff2786bca3b4292b48778235c3

I will updated my patch with some tests in this week.

**Comment by k-keiichi-rh** [comment]
The series of patches are conflicted with 41bb627a2b332e6297f52920f85c0f94bbc04071.
Give me some time to fix the series of patches.

File: pkg/generators/forwarding/fluentd/fluent_conf.go [hunk 1]
```diff
@@ -16,6 +18,45 @@ import (
...
+	if namespaces != nil {
+		namespaceList = strings.Join(namespaces, ",")
+	}
+
+	labelMap, err := metav1.LabelSelectorAsMap(labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("LabelSelector: %v", err)
+        }
+	if labelMap != nil {
+                names := make([]string, 0, len(labelMap))
```
Comment (alanconway) [inline pkg/generators/forwarding/fluentd/fluent_conf.go:40]: Minor: append() works on a nil slice (creates a new slice) so you can do this;
```suggestion
                var names []string
```

**Review by alanconway — APPROVED** [review]
Minor comments on Go idioms, not essential changes.

File: pkg/generators/forwarding/fluentd/fluent_conf.go (hunk 1 above)
Comment (alanconway) [reply pkg/generators/forwarding/fluentd/fluent_conf.go:40]: /lgtm

**Comment by k-keiichi-rh** [comment]
Thanks for your comments. I will fix them.
I have one thing to investigate for the added e2e testcases.
Sometimes tests in input_selector_test.go fail. It looks like the old logs from the previous e2e test remain in FluentdReceiver although a cleanup process like removing /tmp/app.logs, /tmp/app-logs is executed and the FluentdReceiver is recreated every time a new e2e test is executed.
I will summarize my situation with my test logs.

**Comment by alanconway** [comment]
@k-keiichi-rh you need to remove "WIP: " from the title of the PR for it to merge, that will automatically remove the work-in-progress label.

**Review by k-keiichi-rh — COMMENTED** [review]
Sorry for my stupid question.
This issue was caused by an unflushed chunk from the previous testcase.

The following logs are from the testcase just before the failed testcase:
```
{"_ts":"2021-04-22T22:07:28.399164552Z","_file:line":"namespace_filtering_test.go:109","_level":"3","_component":"test","_message":"Running e2e cleanup funct
ions, ","number":11}
...
{"_ts":"2021-04-22T22:08:24.454260913Z","_level":"0","_component":"test","_message":"No pods found for label selection","labels":"test"}
? [SLOW TEST:113.225 seconds]
[Collection] Namespace filtering
/home/cloud/src/github.com/openshift/cluster-logging-operator/test/e2e/collection/fluentd/namespace_filtering_test.go:18
  should send logs from one namespace only
  /home/cloud/src/github.com/openshift/cluster-logging-operator/test/e2e/collection/fluentd/namespace_filtering_test.go:95
```
After finishing namespace_filtering_test.go, the cleanup process is kicked.
And then I checked buffers in /var/lib/fluentd/fluent_receiver after finishing the cleanup process:

```
[root@crc-rsppg-master-0 fluent_receiver]# date; ls /var/lib/fluentd/fluent_receiver; cat /var/lib/fluentd/fluent_receiver/* | strings
Thu Apr 22 22:08:30 UTC 2021
buffer.q5c096e79fd205a98c2700de3de6818d5.log
buffer.q5c096e79fd205a98c2700de3de6818d5.log.meta
buffer.q5c096e7f83399787c1f22dbc722ceb3e.log
buffer.q5c096e7f83399787c1f22dbc722ceb3e.log.meta
docker
container_id
@789e2bb263cdea56e3d2dddb318b7b4a7e06291a091bc5a7c1230f5da18c0c9f
kubernetes
container_name
log-generator
namespace_name
application-ns1
pod_name
log-generator-548446f6dd-6pg6n
container_image
 docker.io/library/busybox:latest
container_image_id
adocker.io/library/busybox@sha256:1ccc0a0ca577e5fb5a0bdf2150a1a9f842f47c8865e861fa0062c5d343eb8cac
pod_id
$e81af5c9-bbe3-42a1-b1e4-69aa799d6595
host
crc-rsppg-master-0
master_url
https://kubernetes.default.svc
namespace_id
$3251f310-336d-4809-b143-1c306d8330d1
flat_labels
component=test
logging-infra=log-generator
pod-template-hash=548446f6dd
provider=openshift
message
47: My life is my message
...
```

Some chunks are left although the previous testcase is completed.
And then a new testcase(input_selector_test.go) is executed:

```
{"_ts":"2021-04-22T22:10:25.199667493Z","_file:line":"input_selector_test.go:232","_level":"3","_component":"test","_message":"Creating ClusterLogForwarder",
"ClusterLogForwarder":"{\"kind\":\"ClusterLogForwarder\",\"apiVersion\":\"logging.openshift.io/v1\",\"metadata\":{\"name\":\"instance\",\"creationTimestamp\"
:null},\"spec\":{\"inputs\":[{\"name\":\"application-logs\",\"application\":{\"namespaces\":[\"application-ns1\"],\"selector\":{\"matchLabels\":{\"env\":\"en
v1\",\"name\":\"app1\"}}}}],\"outputs\":[{\"name\":\"fluent-receiver\",\"type\":\"fluentdForward\",\"url\":\"tcp://fluent-receiver.openshift-logging.svc:2422
4\"}],\"pipelines\":[{\"outputRefs\":[\"fluent-receiver\"],\"inputRefs\":[\"application-logs\"],\"name\":\"fluent-app-logs\"}]},\"status\":{}}"}
{"_ts":"2021-04-22T22:10:25.211193997Z","_level":"0","_component":"test","_message":"NewClusterLogging ","componentTypes":["Collector"]}
...
{"_ts":"2021-04-22T22:10:36.277037348Z","_level":"0","_component":"test","_message":"Print","msg":{"docker":{"container_id":"789e2bb263cdea56e3d2dddb318b7b4a
7e06291a091bc5a7c1230f5da18c0c9f"},"kubernetes":{"container_name":"log-generator","namespace_name":"application-ns1","pod_name":"log-generator-548446f6dd-6pg
6n","container_image":"docker.io/library/busybox:latest","container_image_id":"docker.io/library/busybox@sha256:1ccc0a0ca577e5fb5a0bdf2150a1a9f842f47c8865e86
1fa0062c5d343eb8cac","pod_id":"e81af5c9-bbe3-42a1-b1e4-69aa799d6595","host":"crc-rsppg-master-0","master_url":"https://kubernetes.default.svc","namespace_id"
:"3251f310-336d-4809-b143-1c306d8330d1","flat_labels":["component=test","logging-infra=log-generator","pod-template-hash=548446f6dd","provider=openshift"]},"
message":"53: My life is my message","level":"unknown","hostname":"crc-rsppg-master-0","pipeline_metadata":{"collector":{"ipaddr4":"192.168.126.11","inputnam
e":"fluent-plugin-systemd","name":"fluentd","received_at":"2021-04-22T22:07:32.983291Z","version":"1.7.4 1.6.0"}},"@timestamp":"2021-04-22T22:07:32.36866Z","
viaq_index_name":"app-write","viaq_msg_id":"NGY4MGI2MmItNzU0Mi00ZjdjLTlkM2EtNDhlZmFhYTcxOWVl","systemd":{"t":{},"u":{}},"audit.linux":{},"user":{},"objectRef
":{},"responseStatus":{},"requestReceivedTimestamp":"0001-01-01T00:00:00Z","stageTimestamp":"0001-01-01T00:00:00Z","annotations":{"authorization.k8s.io/decis
ion":"","authorization.k8s.io/reason":""},"openshift":{}}}
```

The chunk from the previous testcase(namespace_filtering_test.go) was flushed and then messages generated by log-generator-548446f6dd-6pg6 which is created by the previous testcase are stored into FluendReceiver.
This message caused the testcase failure.

To avoid this issue, we need to change the order of cleanup like the following:

```
                        AfterEach(func() {
-                               e2e.Cleanup()
                                if _, err = oc.Literal().From("oc delete ns %s", appNamespace1).Run(); err != nil {
                                        Fail("failed to delete namespace")
                                }
                                if _, err = oc.Literal().From("oc delete ns %s", appNamespace2).Run(); err != nil {
                                        Fail("failed to delete namespace")
                                }
-                               e2e.WaitForCleanupCompletion(helpers.OpenshiftLoggingNS, []string{"fluent-receiver", "fluentd"})
                                e2e.WaitForCleanupCompletion(appNamespace1, []string{"test"})
                                e2e.WaitForCleanupCompletion(appNamespace2, []string{"test"})
+                               e2e.Cleanup()
+                               e2e.WaitForCleanupCompletion(helpers.OpenshiftLoggingNS, []string{"fluent-receiver", "fluentd"})
                        }, helpers.DefaultCleanUpTimeout)
```

File: pkg/generators/forwarding/fluentd/fluent_conf.go [hunk 2]
```diff
@@ -16,6 +18,45 @@ import (
...
+	Pipeline        string
+	Namespaces      string
+	Labels          string
+}
+
+func newInputSelectorConf(pipeline string, namespaces []string, labelSelector *metav1.LabelSelector) (*inputSelectorConf, error) {
+	labelList := ""
+	namespaceList := ""
+
+	if namespaces != nil {
```
[the remark this answers is not in the document]
Comment (k-keiichi-rh) [reply pkg/generators/forwarding/fluentd/fluent_conf.go]: I fixed.
```
+func newInputSelectorConf(pipeline string, namespaces []string) (*inputSelectorConf, error) {
+       return &inputSelectorConf{
+               Pipeline: pipeline,
+               Namespaces: strings.Join(namespaces, ","),
+       }, nil
+}
+
```

File: pkg/generators/forwarding/fluentd/fluent_conf.go [hunk 3]
```diff
@@ -16,6 +18,45 @@ import (
...
+		namespaceList = strings.Join(namespaces, ",")
+	}
+
+	labelMap, err := metav1.LabelSelectorAsMap(labelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("LabelSelector: %v", err)
+        }
+	if labelMap != nil {
```
[the remark this answers is not in the document]
Comment (k-keiichi-rh) [reply pkg/generators/forwarding/fluentd/fluent_conf.go]: I fixed.
```
+       labelList := ""
+       var names []string
+
+       labelMap, err := metav1.LabelSelectorAsMap(labelSelector)
+       if err != nil {
+               return nil, fmt.Errorf("LabelSelector: %v", err)
+       }
+       for name := range labelMap {
+               names = append(names, name)
+       }
+       sort.Strings(names)
+       for _, k := range names {
+               if labelList != "" {
+                       labelList += ","
+               }
+               labelList += fmt.Sprintf("%s:%s", k, labelMap[k])
+       }
```

**Review by alanconway — APPROVED** [review]
/lgtm

**Pull request merged**